src/openMVG/multiview/solver_resection_p3p_ding.cpp

```diff
@@ -0,0 +1,365 @@
+// This file is part of OpenMVG, an Open Multiple View Geometry C++ library.
+
+// Copyright (c) 2024 Yaqing Ding
+// Some of the scripts are based on Mark Shachkov (mark.shachkov@gmail.com) and
+// the Lambda-twist P3P implementation Adapted to openMVG by Pierre Moulon
+
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "openMVG/multiview/solver_resection_p3p_ding.hpp"
+#include "openMVG/multiview/projection.hpp"
+
+#include <array>
+
+namespace openMVG {
+namespace euclidean_resection {
+
+/**
+ * @brief Refine a valid solution with a Gauss-Newton Solver.
+ * @param lambda Valid lambdas
+ * @param a12 is the squared distance between X1 and X2
+ * @param a13 is the squared distance between X1 and X3
+ * @param a23 is the squared distance between X2 and X3
+ * @param b12 is the cosine of the angle between bearing vector 1 and bearing
+ * vector 2
+ * @param b13 is the cosine of the angle between bearing vector 1 and bearing
+ * vector 3
+ * @param b23 is the cosine of the angle between bearing vector 2 and bearing
+ * vector 3 The paper note it rarely improve after two iterations. The original
+ * implementation use 5 iterations.
+ */
+static void gauss_newton_refineL(double &lambda1, double &lambda2,
+                                 double &lambda3, const double &a12,
+                                 const double &a13, const double &a23,
+                                 const double &b12, const double &b13,
+                                 const double &b23) {
+  for (int iter = 0; iter < 5; ++iter) {
+    double r1 = (lambda1 * lambda1 - 2.0 * lambda1 * lambda2 * b12 +
+                 lambda2 * lambda2 - a12);
+    double r2 = (lambda1 * lambda1 - 2.0 * lambda1 * lambda3 * b13 +
+                 lambda3 * lambda3 - a13);
+    double r3 = (lambda2 * lambda2 - 2.0 * lambda2 * lambda3 * b23 +
+                 lambda3 * lambda3 - a23);
+    if (std::abs(r1) + std::abs(r2) + std::abs(r3) < 1e-10)
+      return;
+    double x11 = lambda1 - lambda2 * b12;
+    double x12 = lambda2 - lambda1 * b12;
+    double x21 = lambda1 - lambda3 * b13;
+    double x23 = lambda3 - lambda1 * b13;
+    double x32 = lambda2 - lambda3 * b23;
+    double x33 = lambda3 - lambda2 * b23;
+    double detJ = 0.5 / (x11 * x23 * x32 +
+                         x12 * x21 * x33); // half minus inverse determinant
+    // This uses the closed form of the inverse for the jacobean.
+    // Due to the zero elements this actually becomes quite nice.
+    lambda1 += (-x23 * x32 * r1 - x12 * x33 * r2 + x12 * x23 * r3) * detJ;
+    lambda2 += (-x21 * x33 * r1 + x11 * x33 * r2 - x11 * x23 * r3) * detJ;
+    lambda3 += (x21 * x32 * r1 - x11 * x32 * r2 - x12 * x21 * r3) * detJ;
+  }
+};
+
+static inline bool root2real(const double b, const double c, double &r1,
```

Comment:
Some code is most likely identical to https://github.com/openMVG/openMVG/blob/develop/src/openMVG/multiview/solver_resection_p3p_nordberg.cpp#L96
We could maybe factorize in a `p3p_nordberg_utils.hpp` file

Reply:
Moved the repeated scripts into one file.